Misc cleanups

mbouron opened the pull request:
Based on https://github.com/nope-project/nope.gl/pull/71.

ubitux commented on libnopegl/src/backends/gl/pipeline_gl.c:
```diff
@@ -375,22 +375,11 @@ int ngli_pipeline_gl_init(struct pipeline *s, const struct pipeline_params *para
 {
     struct pipeline_gl *s_priv = (struct pipeline_gl *)s;
 
-    s->type     = params->type;
```
initilization / initialization typo in commit message

mbouron replied:
Fixed.

ubitux commented on libnopegl/src/rendertarget.c:
```diff
@@ -38,6 +39,10 @@ int ngli_rendertarget_init(struct rendertarget *s, const struct rendertarget_par
     ngli_assert(params->nb_colors <= limits->max_color_attachments);
 
+    if (params->depth_stencil.resolve_target) {
+        ngli_assert(features & NGLI_FEATURE_DEPTH_STENCIL_RESOLVE);
```
There is a features parameter exposed to the rtt node: I'm assuming it can't cause this assert to be triggered?

mbouron replied:
Yes, see:
``` c
        if (!(gpu_ctx->features & NGLI_FEATURE_DEPTH_STENCIL_RESOLVE) && o->samples > 0) {
            LOG(ERROR, "context does not support resolving depth/stencil attachments");
            return NGL_ERROR_GRAPHICS_UNSUPPORTED;
        }
```

ubitux commented on libnopegl/src/backends/vk/rendertarget_vk.c:
```diff
@@ -55,22 +55,23 @@ static VkAttachmentStoreOp get_vk_store_op(int store_op)
+#define ARRAY(type, name, size) type name[(size)] = {0}; size_t nb_##name = 0
+
+#define ARRAY_APPEND(name, ...) do { \
+    name[nb_##name] = __VA_ARGS__;   \
```
Why a va-arg here? 🤔

mbouron replied:
To allow the following use:
``` c
        ARRAY_APPEND(color_refs, (VkAttachmentReference) {
            .attachment = (uint32_t)nb_descs,
            .layout     = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
        };
```
Without VA_ARGS you get a compile error as:
``` c
(VkAttachmentReference) {
            .attachment = (uint32_t)nb_descs,
            .layout     = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
        };
```
is treated as multiple arguments.

ubitux commented on libnopegl/src/backends/vk/rendertarget_vk.c:
```diff
@@ -87,7 +88,7 @@ static VkResult vk_create_compatible_renderpass(struct gpu_ctx *s, const struct
         const VkAttachmentLoadOp load_op   = params ? get_vk_load_op(params->colors[i].load_op)   : VK_ATTACHMENT_LOAD_OP_DONT_CARE;
         const VkAttachmentStoreOp store_op = params ? get_vk_store_op(params->colors[i].store_op) : VK_ATTACHMENT_STORE_OP_DONT_CARE;
-        descs[nb_descs] = (VkAttachmentDescription) {
+        ARRAY_APPEND(descs, (VkAttachmentDescription) {
```
I don't find this to be an improvement in readability tbh. I'd rather do `desc[nb_descs++] = ...` so that the two variables are not decorrelated. The switch to `size_t` is good though (and could probably be split out)

mbouron replied:
Switched to size_t in a separate commit.

mbouron replied:
The commit has been dropped.

mbouron commented on libnopegl/src/backends/gl/gpu_ctx_gl.c:
```diff
@@ -984,7 +984,7 @@ const struct gpu_ctx_class ngli_gpu_ctx_##cls_suffix = {
                                                                                  \
     .get_default_rendertarget           = gl_get_default_rendertarget,           \
-    .get_default_rendertarget_desc      = gl_get_default_rendertarget_desc,      \
+    .get_default_rendertarget_layout      = gl_get_default_rendertarget_layout,      \
```
Fixed.

mbouron commented on libnopegl/src/backends/vk/gpu_ctx_vk.c:
```diff
@@ -1528,7 +1528,7 @@ const struct gpu_ctx_class ngli_gpu_ctx_vk = {
 
     .get_default_rendertarget           = vk_get_default_rendertarget,
-    .get_default_rendertarget_desc      = vk_get_default_rendertarget_desc,
+    .get_default_rendertarget_layout      = vk_get_default_rendertarget_layout,
```
Fixed.

ubitux replied:
Seems you forgot this one

mbouron replied:
Fixed.

mbouron commented on libnopegl/src/rendertarget.h:
```diff
@@ -37,16 +37,16 @@ enum {
 };
 
-struct attachment_desc {
+struct attachment_layout_entry {
```
Fixed. It is indeed `rendertarget_layout_entry`.

ubitux commented on libnopegl/src/backends/vk/rendertarget_vk.c:
```diff
@@ -55,22 +55,23 @@ static VkAttachmentStoreOp get_vk_store_op(int store_op)
-
-    VkAttachmentReference resolve_refs[NGLI_MAX_COLOR_ATTACHMENTS + 1] = {0};
-    int nb_resolve_refs = 0;
+    ARRAY(VkAttachmentDescription, descs, NGLI_MAX_COLOR_ATTACHMENTS + 1);
```
what happened to the `2*`?

mbouron replied:
This commit has been dropped.

ubitux commented on libnopegl/src/node_text.c:
```diff
@@ -482,7 +482,7 @@ static int init_subdesc(struct ngl_node *node,
             .topology     = NGLI_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP,
             .state        = *graphics_state,
-            .rt_desc      = rnode->rendertarget_desc,
+            .rt_layout      = rnode->rendertarget_layout,
```
style: there is a bunch of small spacing alignment issues in this commit due to `rt_desc` to `rt_layout` rename

ubitux reviewed (approved):
Aside from the small style issue LGTM

mbouron commented:
> Aside from the small style issue LGTM

I added the following extra commit (that fixes the changes you reviewed with the Vulkan backend):
```
    vk: use common ngli graphics calls where appropriate
    
    In the upcoming commits, common ngli graphics calls will handle more
    non-backend specific logic, for example: initializing common fields,
    validating parameters and states.
    
    The downside of this approach is that we currently lose some error
    information in the gpu_ctx_vk and texture_vk modules. In practice, this
    is not an issue for the gpu_ctx_vk module as the error is ultimately
    converted to the corresponding ngl ret code (without any logging). For
    the texture_vk module and the Vulkan codebase in general, we need to
    figure out a proper way to report Vulkan errors.
```

ubitux commented on libnopegl/src/backends/vk/gpu_ctx_vk.c:
```diff
@@ -66,26 +66,30 @@ static VkResult create_dummy_texture(struct gpu_ctx *s)
-        return res;
+    int ret = ngli_texture_init(s_priv->dummy_texture, &params);
+    if (ret < 0)
+        return VK_ERROR_UNKNOWN;
```
Afaict everything except the swapchain could work with ngl retcode directly with no loss of information. Can be done later™.

mbouron replied:
I think I'll do that in a separate PR with more cleanups.
